drivers/sensor/st/lsm6dsv16x/lsm6dsv16x_decoder.c

```diff
@@ -0,0 +1,446 @@
+ * GAIN_UNIT_G is expressed in udps/LSB.
+ */
+static const int32_t gyro_scaler[] = {
+	[LSM6DSV16X_DT_FS_125DPS] = SENSOR_SCALE_UDPS_TO_URADS(GAIN_UNIT_G),
+	[LSM6DSV16X_DT_FS_250DPS] = SENSOR_SCALE_UDPS_TO_URADS(2 * GAIN_UNIT_G),
+	[LSM6DSV16X_DT_FS_500DPS] = SENSOR_SCALE_UDPS_TO_URADS(4 * GAIN_UNIT_G),
+	[LSM6DSV16X_DT_FS_1000DPS] = SENSOR_SCALE_UDPS_TO_URADS(8 * GAIN_UNIT_G),
+	[LSM6DSV16X_DT_FS_2000DPS] = SENSOR_SCALE_UDPS_TO_URADS(16 * GAIN_UNIT_G),
+	[LSM6DSV16X_DT_FS_4000DPS] = SENSOR_SCALE_UDPS_TO_URADS(32 * GAIN_UNIT_G),
+};
+
+static int lsm6dsv16x_decoder_get_frame_count(const uint8_t *buffer,
+					      struct sensor_chan_spec chan_spec,
+					      uint16_t *frame_count)
+{
+	struct lsm6dsv16x_fifo_data *data = (struct lsm6dsv16x_fifo_data *)buffer;
+	const struct lsm6dsv16x_decoder_header *header = &data->header;
+
+	if (chan_spec.chan_idx != 0) {
+		return -ENOTSUP;
+	}
+
+	if (!header->is_fifo) {
+		switch (chan_spec.chan_type) {
+		case SENSOR_CHAN_ACCEL_X:
+		case SENSOR_CHAN_ACCEL_Y:
+		case SENSOR_CHAN_ACCEL_Z:
+		case SENSOR_CHAN_ACCEL_XYZ:
+		case SENSOR_CHAN_GYRO_X:
+		case SENSOR_CHAN_GYRO_Y:
+		case SENSOR_CHAN_GYRO_Z:
+		case SENSOR_CHAN_GYRO_XYZ:
+		case SENSOR_CHAN_DIE_TEMP:
+			*frame_count = 1;
+			return 0;
+		default:
+			return -ENOTSUP;
+		}
+
+		return 0;
+	}
+
+#ifdef CONFIG_LSM6DSV16X_STREAM
+	*frame_count = data->fifo_count;
+#endif
+	return 0;
+}
+
+#ifdef CONFIG_LSM6DSV16X_STREAM
+static int lsm6dsv16x_decode_fifo(const uint8_t *buffer, struct sensor_chan_spec chan_spec,
+				  uint32_t *fit, uint16_t max_count, void *data_out)
+{
+	const struct lsm6dsv16x_fifo_data *edata = (const struct lsm6dsv16x_fifo_data *)buffer;
+	const uint8_t *buffer_end;
+	const struct lsm6dsv16x_decoder_header *header = &edata->header;
+	int count = 0;
+	uint8_t fifo_tag;
+	uint16_t xl_count = 0, gy_count = 0;
+
+#if defined(CONFIG_LSM6DSV16X_ENABLE_TEMP)
+	uint16_t temp_count = 0;
+#endif
+
+	buffer += sizeof(struct lsm6dsv16x_fifo_data);
+	buffer_end = buffer + LSM6DSV16X_FIFO_SIZE(edata->fifo_count);
+
+	((struct sensor_data_header *)data_out)->base_timestamp_ns = edata->header.timestamp;
+
+	while (count < max_count && buffer < buffer_end) {
+		const uint8_t *frame_end = buffer;
+		uint8_t skip_frame;
+
+		skip_frame = 0;
+		frame_end += LSM6DSV16X_FIFO_ITEM_LEN;
+
+		fifo_tag = (buffer[0] >> 3);
+
+		switch (fifo_tag) {
+		case LSM6DSV16X_XL_NC_TAG: {
+			struct sensor_three_axis_data *out = data_out;
+			int16_t x, y, z;
+			const int32_t scale = accel_scaler[header->accel_fs];
+
+			xl_count++;
+			if ((uintptr_t)buffer < *fit) {
+				/* This frame was already decoded, move on to the next frame */
+				buffer = frame_end;
+				continue;
+			}
+
+			if (!SENSOR_CHANNEL_IS_ACCEL(chan_spec.chan_type)) {
+				buffer = frame_end;
+				continue;
+			}
+
+			out->header.base_timestamp_ns = edata->header.timestamp;
+			out->readings[count].timestamp_delta =
+				(xl_count - 1) * accel_period_ns[edata->accel_batch_odr];
```

Comment:
I'm rather confused by this... but I see other drivers do this as well.

If the timestamp is recorded when the fifo is full or the watermark is hit, then that time is the time of the very last sample in the fifo.... but as it is a fifo, the first sample read out will be the oldest sample which the actual time of the sampling would of been quite a bit before the time of the last sampling.

In my implementation... I did this
```c
static int lsm6dsv16x_fifo_decode(const uint8_t *buffer, struct sensor_chan_spec chan_spec,
				  uint32_t *fit, uint16_t max_count, void *data_out)
{
	const struct lsm6dsv16x_fifo_data *edata = (const struct lsm6dsv16x_fifo_data *)buffer;
	const struct lsm6dsv16x_decoder_header *header = &edata->header;
	struct lsm6dsv16x_settings cfg = {
		.accel_range = edata->header.accel_fs,
		.gyro_range = edata->header.gyro_fs,
	};
	const uint8_t *buffer_end =
		buffer + sizeof(struct lsm6dsv16x_fifo_data) + (edata->fifo_count * 7);
	const uint8_t *temp_buffer = buffer;
	lsm6dsv16x_fifo_data_out_tag_t tag;
	int count = 0;
	int accel_frame_count = 0;
	int gyro_frame_count = 0;
	int total_accel_frames = 0;
	int total_gyro_frames = 0;
	int rc;

	if ((uintptr_t)buffer_end <= *fit || chan_spec.chan_idx != 0) {
		return 0;
	}

	while (temp_buffer < buffer_end) {
		memcpy(&tag, &buffer[0], sizeof(tag));
		/* frame counters */
		switch (tag.tag_sensor) {
		case LSM6DSV16X_XL_NC_TAG:
			total_accel_frames++;
			break;
		case LSM6DSV16X_GY_NC_TAG:
			total_gyro_frames++;
			break;
		case LSM6DSV16X_CFG_CHANGE_TAG:
			/* TODO: record ODR changes */
			break;
		default:
			break;
		}

		/* A frame is 7 Bytes */
		temp_buffer += 7;
	}

	/*
	 * as the timestamp is when the fifo event occured, the base timestamp needs the time of the
	 * first frame which will be the first frame read out of the FIFO. Back date the timestamp by
	 * the total number of frames in the FIFO * the period of the ODR.
	 * TODO: This could have issues of the ODR was changed while streaming.
	 */
	if (IS_ACCEL(chan_spec.chan_type)) {
		((struct sensor_data_header *)data_out)->base_timestamp_ns =
			edata->header.timestamp -
			(total_accel_frames - 1) * period_ns[edata->accel_odr];
	} else if (IS_GYRO(chan_spec.chan_type)) {
		((struct sensor_data_header *)data_out)->base_timestamp_ns =
			edata->header.timestamp -
			(total_gyro_frames - 1) * period_ns[edata->gyro_odr];
	}

	/* Get the start of the FIFO data */
	buffer += sizeof(struct lsm6dsv16x_fifo_data);
	while (count < max_count && buffer < buffer_end) {
		memcpy(&tag, &buffer[0], sizeof(tag));
		/* frame counters */
		switch (tag.tag_sensor) {
		case LSM6DSV16X_XL_NC_TAG:
			accel_frame_count++;
			break;
		case LSM6DSV16X_GY_NC_TAG:
			gyro_frame_count++;
			break;
		default:
			break;
		}

		if ((uintptr_t)buffer < *fit) {
			/* This frame was already decoded, move on to the next frame */
			buffer += 7;
			continue;
		}

		switch (tag.tag_sensor) {
		case LSM6DSV16X_XL_NC_TAG:
			if (IS_ACCEL(chan_spec.chan_type)) {
				/* Decode accel */
				struct sensor_three_axis_data *data =
					(struct sensor_three_axis_data *)data_out;
				uint64_t accel_period_ns = period_ns[edata->accel_odr];

				rc = lsm6dsv16x_get_shift(SENSOR_CHAN_ACCEL_XYZ, header->accel_fs,
							  header->gyro_fs, &data->shift);
				if (rc != 0) {
					return -EINVAL;
				}

				/* extract raw data from fifo frame */
				int16_t accel_x_raw = sys_get_le16(&buffer[1]);
				int16_t accel_y_raw = sys_get_le16(&buffer[3]);
				int16_t accel_z_raw = sys_get_le16(&buffer[5]);

				data->readings[count].timestamp_delta =
					(accel_frame_count - 1) * accel_period_ns;
				lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_ACCEL_X,
							      accel_x_raw,
							      &data->readings[count].x);
				lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_ACCEL_Y,
							      accel_y_raw,
							      &data->readings[count].y);
				lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_ACCEL_Z,
							      accel_z_raw,
							      &data->readings[count].z);
				count++;
			}
			break;
		case LSM6DSV16X_GY_NC_TAG:
			if (IS_GYRO(chan_spec.chan_type)) {
				/* Decode gyro */
				struct sensor_three_axis_data *data =
					(struct sensor_three_axis_data *)data_out;
				uint64_t gyro_period_ns = period_ns[edata->gyro_odr];

				rc = lsm6dsv16x_get_shift(SENSOR_CHAN_GYRO_XYZ, header->accel_fs,
							  header->gyro_fs, &data->shift);
				if (rc != 0) {
					return -EINVAL;
				}

				/* extract raw data from fifo frame */
				int16_t gyro_x_raw = sys_get_le16(&buffer[1]);
				int16_t gyro_y_raw = sys_get_le16(&buffer[3]);
				int16_t gyro_z_raw = sys_get_le16(&buffer[5]);

				data->readings[count].timestamp_delta =
					(gyro_frame_count - 1) * gyro_period_ns;
				lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_GYRO_X, gyro_x_raw,
							      &data->readings[count].x);
				lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_GYRO_Y, gyro_y_raw,
							      &data->readings[count].y);
				lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_GYRO_Z, gyro_z_raw,
							      &data->readings[count].z);
				count++;
			}
			break;
#if defined(CONFIG_LSM6DSV16X_ENABLE_TEMP)
		case LSM6DSV16X_TEMPERATURE_TAG:
			if (chan_spec.chan_type == SENSOR_CHAN_DIE_TEMP) {
				/* Decode temperature */
				struct sensor_q31_data *data = (struct sensor_q31_data *)data_out;

				rc = lsm6dsv16x_get_shift(SENSOR_CHAN_DIE_TEMP, 0, &data->shift);
				if (rc != 0) {
					return -EINVAL;
				}

				/* extract raw data from fifo frame */
				int16_t temp_raw = sys_get_le16(&buffer[1]);

				lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_DIE_TEMP, temp_raw,
							      &data->readings[count].temperature);
				count++;
			}
			break;
#endif /* CONFIG_LSM6DSV16X_ENABLE_TEMP */
		case LSM6DSV16X_TIMESTAMP_TAG:
			/* Decode timestamp */
			// uint32_t timestamp = sys_get_le32(&buffer[1]);

			/* TODO: Word also contains other metadata that may be useful */

			break;
		case LSM6DSV16X_FIFO_EMPTY:
		default:
			break;
		}
		/* A frame is 7 Bytes */
		buffer += 7;
		*fit = (uintptr_t)buffer;
	}
	return count;
}

static int lsm6dsv16x_one_shot_decode(const uint8_t *buffer, struct sensor_chan_spec chan_spec,
				      uint32_t *fit, uint16_t max_count, void *data_out)
{
	const struct lsm6dsv16x_encoded_data *edata =
		(const struct lsm6dsv16x_encoded_data *)buffer;
	const struct lsm6dsv16x_decoder_header *header = &edata->header;
	struct lsm6dsv16x_settings cfg = {
		.accel_range = edata->header.accel_fs,
		.gyro_range = edata->header.gyro_fs,
	};
	int rc;

	if (*fit != 0) {
		return 0;
	}
	if (max_count == 0 || chan_spec.chan_idx != 0) {
		return -EINVAL;
	}

	switch (chan_spec.chan_type) {
	case SENSOR_CHAN_ACCEL_X:
	case SENSOR_CHAN_ACCEL_Y:
	case SENSOR_CHAN_ACCEL_Z:
	case SENSOR_CHAN_ACCEL_XYZ: {
		if (!edata->has_accel) {
			return -ENODATA;
		}

		struct sensor_three_axis_data *out = (struct sensor_three_axis_data *)data_out;

		out->header.base_timestamp_ns = edata->header.timestamp;
		out->header.reading_count = 1;
		rc = lsm6dsv16x_get_shift(SENSOR_CHAN_ACCEL_XYZ, header->accel_fs, header->gyro_fs,
					  &out->shift);
		if (rc != 0) {
			return -EINVAL;
		}

		lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_ACCEL_X, edata->accel_xyz[0],
					      &out->readings[0].x);
		lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_ACCEL_Y, edata->accel_xyz[1],
					      &out->readings[0].y);
		lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_ACCEL_Z, edata->accel_xyz[2],
					      &out->readings[0].z);

		*fit = 1;
		return 1;
	}
	case SENSOR_CHAN_GYRO_X:
	case SENSOR_CHAN_GYRO_Y:
	case SENSOR_CHAN_GYRO_Z:
	case SENSOR_CHAN_GYRO_XYZ: {
		if (!edata->has_gyro) {
			return -ENODATA;
		}

		struct sensor_three_axis_data *out = (struct sensor_three_axis_data *)data_out;

		out->header.base_timestamp_ns = edata->header.timestamp;
		out->header.reading_count = 1;
		rc = lsm6dsv16x_get_shift(SENSOR_CHAN_GYRO_XYZ, header->accel_fs, header->gyro_fs,
					  &out->shift);
		if (rc != 0) {
			return -EINVAL;
		}

		lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_GYRO_X, edata->accel_xyz[0],
					      &out->readings[0].x);
		lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_GYRO_Y, edata->accel_xyz[1],
					      &out->readings[0].y);
		lsm6dsv16x_convert_raw_to_q31(&cfg, SENSOR_CHAN_GYRO_Z, edata->accel_xyz[2],
					      &out->readings[0].z);

		*fit = 1;
		return 1;
	}
#ifdef CONFIG_LSM6DSV16X_TEMPERATURE
	case SENSOR_CHAN_DIE_TEMP: {
		if (!edata->has_temp) {
			return -ENODATA;
		}

		struct sensor_q31_data *out = (struct sensor_q31_data *)data_out;

		out->header.base_timestamp_ns = edata->header.timestamp;
		out->header.reading_count = 1;
		rc = lsm6dsv16x_get_shift(SENSOR_CHAN_DIE_TEMP, 0, &out->shift);
		if (rc != 0) {
			return -EINVAL;
		}

		lsm6dsv16x_convert_raw_temp_to_q31(edata->temp, &out->readings[0].temperature);

		*fit = 1;
		return 1;
	}
#endif /* CONFIG_LSM6DSV16X_TEMPERATURE */
	default:
		return -EINVAL;
	}
}
```

Reply:
Leave it open for the moment... I'll think more about it

Reply:
think you are right. I'll count the frames and set base_timestamp_ns accordingly